B4---C++-Programming/cpp_bomberman/annexes/doxygen/man/man3/GenerateMaze.3
.TH "GenerateMaze" 3 "Tue Jun 9 2015" "Version 0.53" "cpp_bomberman" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GenerateMaze \- Generate maze (char *)  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <GenerateMaze\&.hh>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBcreate\fP (int width, int height)"
.br
.RI "\fIGenerate normal map (like bomberman) \fP"
.ti -1c
.RI "void \fBcreateRandom\fP (int width, int height)"
.br
.RI "\fIGenerate random map\&. \fP"
.ti -1c
.RI "char * \fBgetMap\fP () const "
.br
.RI "\fIget the map generated \fP"
.ti -1c
.RI "std::vector< std::pair< int, int > > & \fBgetEmptyCells\fP ()"
.br
.RI "\fIget empty cells in map \fP"
.ti -1c
.RI "void \fBsetCell\fP (int x, int y, int id_cell)"
.br
.RI "\fIset a cell in the map \fP"
.ti -1c
.RI "int \fBgetCell\fP (int x, int y)"
.br
.RI "\fIget a cell in the map \fP"
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBGenerateMaze\fP & \fBgetInstance\fP ()"
.br
.RI "\fIUsed to return the current instance\&. \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "\fBGenerateMaze\fP & \fBoperator=\fP (const \fBGenerateMaze\fP &)"
.br
.RI "\fIprivate = operator to unable it \fP"
.ti -1c
.RI "\fBGenerateMaze\fP (const \fBGenerateMaze\fP &)"
.br
.RI "\fIprivate copy constructor \fP"
.ti -1c
.RI "\fBGenerateMaze\fP ()"
.br
.RI "\fI\fBGenerateMaze\fP's constructor private\&. \fP"
.ti -1c
.RI "\fB~GenerateMaze\fP ()"
.br
.RI "\fI\fBGenerateMaze\fP's destructor private\&. \fP"
.ti -1c
.RI "void \fBinit\fP (int width, int height)"
.br
.RI "\fIinit map for create \fP"
.ti -1c
.RI "void \fBinitRandom\fP (int width, int height)"
.br
.RI "\fIinit map for createRandom \fP"
.ti -1c
.RI "int \fBrandomDirection\fP () const "
.br
.RI "\fIrandom in 0 - 3 (included) \fP"
.ti -1c
.RI "bool \fBsetWall\fP (int x, int y, int xModifier, int yModifier)"
.br
.RI "\fIintern function for createRandom \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "int \fBm_width\fP"
.br
.ti -1c
.RI "int \fBm_height\fP"
.br
.ti -1c
.RI "char * \fBm_maze\fP"
.br
.ti -1c
.RI "std::vector< std::pair< int, int > > \fBm_emptyCell\fP"
.br
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static \fBGenerateMaze\fP \fBm_instance\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Generate maze (char *) 

Singleton to create a specific map 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "GenerateMaze::GenerateMaze (const \fBGenerateMaze\fP &)\fC [inline]\fP, \fC [private]\fP"

.PP
private copy constructor Unused 
.SS "GenerateMaze::GenerateMaze ()\fC [private]\fP"

.PP
\fBGenerateMaze\fP's constructor private\&. Used to create a static instance (singleton) We will set default value for width, height and maze\&. 
.SS "GenerateMaze::~GenerateMaze ()\fC [private]\fP"

.PP
\fBGenerateMaze\fP's destructor private\&. Used to destroy instance\&. It will delete the maze\&. 
.SH "Member Function Documentation"
.PP 
.SS "void GenerateMaze::create (int width, int height)"

.PP
Generate normal map (like bomberman) It will call the init private function and generate a map
.PP
\fBParameters:\fP
.RS 4
\fIwidth\fP the maze's width 
.br
\fIheigth\fP the maze's height 
.RE
.PP

.SS "void GenerateMaze::createRandom (int width, int height)"

.PP
Generate random map\&. It will call create random map with specific width / height
.PP
\fBParameters:\fP
.RS 4
\fIwidth\fP the maze's width 
.br
\fIheight\fP the maze's height 
.RE
.PP

.SS "int GenerateMaze::getCell (int x, int y)"

.PP
get a cell in the map Get a cell with specific x and y position
.PP
\fBParameters:\fP
.RS 4
\fIx\fP the x's position in map we want to get 
.br
\fIy\fP the y's position in map we want to get 
.RE
.PP
\fBReturns:\fP
.RS 4
Get the cell's id like wall, empty etc 
.RE
.PP

.SS "std::vector<std::pair <int, int> >& GenerateMaze::getEmptyCells ()"

.PP
get empty cells in map Get the empty cells in the map generated
.PP
\fBReturns:\fP
.RS 4
Cells' vector with empty's position 
.RE
.PP

.SS "static \fBGenerateMaze\fP& GenerateMaze::getInstance ()\fC [static]\fP"

.PP
Used to return the current instance\&. Our class is a singleton so we need a function to use it
.PP
\fBReturns:\fP
.RS 4
the current class's instance 
.RE
.PP

.SS "char* GenerateMaze::getMap () const"

.PP
get the map generated Get the map generated by create or createRandom Or null if no map generated
.PP
\fBReturns:\fP
.RS 4
Current map generated 
.RE
.PP

.SS "void GenerateMaze::init (int width, int height)\fC [private]\fP"

.PP
init map for create Init map with specific pattern (like bomberman)
.PP
\fBParameters:\fP
.RS 4
\fIwidth\fP maze's width size 
.br
\fIheight\fP maze's height size 
.RE
.PP

.SS "void GenerateMaze::initRandom (int width, int height)\fC [private]\fP"

.PP
init map for createRandom init map for createRandom with ID_WALL or ID_BLOCK
.PP
\fBParameters:\fP
.RS 4
\fIwidth\fP maze's width size 
.br
\fIheight\fP maze's height size 
.RE
.PP

.SS "\fBGenerateMaze\fP& GenerateMaze::operator= (const \fBGenerateMaze\fP &)\fC [private]\fP"

.PP
private = operator to unable it Unused and undefined 
.SS "int GenerateMaze::randomDirection () const\fC [private]\fP"

.PP
random in 0 - 3 (included) It will generate with random_device, mt19937 and uniform_int_distribution a number in 0 - 3 (included)
.PP
\fBReturns:\fP
.RS 4
A number in 0 - 3 (included) 
.RE
.PP

.SS "void GenerateMaze::setCell (int x, int y, int id_cell)"

.PP
set a cell in the map it will a set a cell in the map with an id
.PP
\fBParameters:\fP
.RS 4
\fIx\fP the x's position in map we want to set 
.br
\fIy\fP the y's position in map we want to set 
.br
\fIid_cell\fP set the cell with specific id (like wall etc) 
.RE
.PP

.SS "bool GenerateMaze::setWall (int x, int y, int xModifier, int yModifier)\fC [private]\fP"

.PP
intern function for createRandom Check if the new position is a valid one in createRandom's algo
.PP
current x's position in the map  current y's position in the map  modifier we want to apply on our current position (like 1, -1, \&.\&.\&.)  modifier we want to apply on our current position (like 1, -1, \&.\&.\&.) 
.PP
\fBReturns:\fP
.RS 4
true if the new position isn't valid else false 
.RE
.PP

.SH "Member Data Documentation"
.PP 
.SS "std::vector<std::pair <int, int> > GenerateMaze::m_emptyCell\fC [private]\fP"
used to increase cell research 
.SS "int GenerateMaze::m_height\fC [private]\fP"
maze's height we will generate 
.SS "\fBGenerateMaze\fP GenerateMaze::m_instance\fC [static]\fP, \fC [private]\fP"
used in singleton 
.SS "char* GenerateMaze::m_maze\fC [private]\fP"
the maze generated will be hold here 
.SS "int GenerateMaze::m_width\fC [private]\fP"
maze's width we will generate 

.SH "Author"
.PP 
Generated automatically by Doxygen for cpp_bomberman from the source code\&.


B4---C++-Programming/cpp_bomberman/src/others/GenerateMaze.cpp
//
// Created by ankirama on 5/12/15.
//

#include "GenerateMaze.hh"

static std::random_device   _rd;
static std::mt19937         _gen(_rd());

GenerateMaze    GenerateMaze::m_instance = GenerateMaze();

GenerateMaze::GenerateMaze() {
  this->m_width = 0;
  this->m_height = 0;
  this->m_maze = 0;
}

GenerateMaze::~GenerateMaze() {
    if (this->m_maze)
        delete this->m_maze;
}

void					GenerateMaze::init(int width, int height) {
  this->m_width = width + 2;
  this->m_height = height + 2;
  if (this->m_maze)
    delete this->m_maze;
  this->m_maze = new char[this->m_width * this->m_height];
  for (int i = 0; i < this->m_width; ++i) {
    for (int j = 0; j < this->m_height; ++j) {
      if (i == 0 || i == (this->m_width - 1) || j == 0 || j == (this->m_height - 1))
	  this->m_maze[j * this->m_width + i] = ID_WALL;
      else if (i % 2 == 0 && j % 2 == 0)
	this->m_maze[j * this->m_width + i] = ID_WALL;
      else
	this->m_maze[j * this->m_width + i] = ID_BLOCK;
    }
  }
}

void					GenerateMaze::initRandom(int width, int height) {
  this->m_width = width;
  this->m_height = height;
  if (this->m_maze)
    delete this->m_maze;
  this->m_maze = new char[this->m_width * this->m_height];
  for (int i = 0; i < this->m_width * this->m_height; ++i) {
    if (i % this->m_width == 0 || i < this->m_width ||
	(i % this->m_width == this->m_width - 1) || i > (this->m_height * this->m_width - this->m_width))
      this->m_maze[i] = ID_WALL;
    else if (this->randomDirection() % 2 == 0) {
      this->m_maze[i] = ID_WALL;
    }
    else {
      this->m_maze[i] = ID_BLOCK;
    }
  }
}

bool					GenerateMaze::setWall(int x, int y, int xModifier, int yModifier) {
  if (x + xModifier < 0 || y + yModifier < 0 ||
    x + xModifier >= this->m_width ||
    y + yModifier >= this->m_height ||
    this->m_maze[(y + yModifier) * this->m_width + x + xModifier] == ID_EMPTY)
    return true;
  else {
    this->m_maze[(y + yModifier) * this->m_width + x + xModifier] = ID_EMPTY;
    return false;
  }
}

void                    GenerateMaze::create(int width, int height) {
  this->init(width, height);
}

void					GenerateMaze::createRandom(int width, int height) {
  std::list <std::pair <int, int> >	openWalls;
  bool					isRemoveWall;

  this->initRandom(width, height);
  openWalls.push_back(std::make_pair(this->m_width / 2, this->m_height / 2));
  while (!openWalls.empty()) {
    std::list <std::pair <int, int> >::iterator tmp;
    tmp = openWalls.begin();
    while (tmp != openWalls.end()) {
      isRemoveWall = false;
      switch (this->randomDirection()) {
      case 0:
	    isRemoveWall = this->setWall(tmp->first, tmp->second -= 2, 0, 1);
        break;
      case 1:
	    isRemoveWall = this->setWall(tmp->first, tmp->second += 2, 0, -1);
	    break;
      case 2:
	    isRemoveWall = this->setWall(tmp->first -= 2, tmp->second, 1, 0);
	    break;
      case 3:
	    isRemoveWall = this->setWall(tmp->first += 2, tmp->second, -1, 0);
	    break;
      }
      if (isRemoveWall)
	    tmp = openWalls.erase(tmp);
      else {
	    openWalls.push_back(std::make_pair(tmp->first, tmp->second));
        if (this->randomDirection() % 2 == 0)
	      openWalls.push_back(std::make_pair(tmp->first, tmp->second));
        this->m_emptyCell.push_back(std::make_pair(tmp->first, tmp->second));
        this->m_maze[tmp->second * this->m_width + tmp->first] = ID_EMPTY;
	    ++tmp;
      }
    }
  }
}

std::vector <std::pair <int, int> >& GenerateMaze::getEmptyCells() {
  return this->m_emptyCell;
}

void GenerateMaze::setCell(int x, int y, int id_cell) {
  if (this->m_maze[y * this->m_width + x] == ID_EMPTY) {
    std::pair<int, int> tmp = std::make_pair(x, y);
    unsigned int i = 0;
    while (i < this->m_emptyCell.size() &&
            !(this->m_emptyCell[i].first == tmp.first &&
                    this->m_emptyCell[i].second == tmp.second))
      ++i;
    if (i < this->m_emptyCell.size())
      this->m_emptyCell.erase(this->m_emptyCell.begin() + i);
  }
  this->m_maze[y * this->m_width + x] = id_cell;
  if (id_cell == ID_EMPTY)
    this->m_emptyCell.push_back(std::make_pair(x, y));
}

int GenerateMaze::getCell(int x, int y) {
  return this->m_maze[y * this->m_width + x];
}

char           *GenerateMaze::getMap() const {
    return this->m_maze;
}

int                     GenerateMaze::randomDirection() const {
    static std::uniform_int_distribution<> disCell(0, 3);
    return static_cast<int>(disCell(_gen));
}

GenerateMaze&   GenerateMaze::getInstance() {
    return m_instance;
}
